src/performance/metrics/StatisticsCollector.cpp

```diff
@@ -12,63 +12,333 @@ limitations under the License.
-    FILE *file = fopen("/proc/self/status", "r");
-    long result = -1;
-    char line[LINE_BUF_SIZE];
+static double measureProcStatRate(const char* prefix, int prefixLen, int sleepSeconds) {
+    // First reading
+    long firstValue = readProcStatValue(prefix, prefixLen);
+    if (firstValue == -1) {
+        stat_logger.error(std::string("Could not read initial ") + prefix + " value");
+        return -1.0;
+    }
 
-    while (fgets(line, LINE_BUF_SIZE, file) != NULL) {
-        if (strncmp(line, "VmSize:", 7) == 0) {
-            result = parseLine(line);
-            break;
+    // Record start time
+    struct timespec startTime;
+    struct timespec endTime;
+    clock_gettime(CLOCK_MONOTONIC, &startTime);
+
+    // Sleep for measurement interval
+    sleep(sleepSeconds);
+
+    // Second reading
+    long secondValue = readProcStatValue(prefix, prefixLen);
+    if (secondValue == -1) {
+        stat_logger.error(std::string("Could not read final ") + prefix + " value");
+        return -1.0;
+    }
+
+    // Record end time
+    clock_gettime(CLOCK_MONOTONIC, &endTime);
+
+    // Calculate elapsed time in seconds
+    double elapsedTime = calculateElapsedTime(startTime, endTime);
+
+    if (elapsedTime <= 0.0) {
+        stat_logger.error(std::string("Invalid elapsed time for ") + prefix + " calculation");
+        return -1.0;
+    }
+
+    // Calculate rate per second
+    long valueDiff = secondValue - firstValue;
+    if (valueDiff < 0) {
+        stat_logger.error(std::string(prefix) + " counter wrapped or invalid");
+        return -1.0;
+    }
+
+    return (double)valueDiff / elapsedTime;
+}
+
+// CPU statistics operations
+static bool readCpuStats(std::vector<std::vector<long>> &readings, const std::string& errorContext = "") {
+    std::ifstream file("/proc/stat");
+    if (!file.is_open()) {
+        std::string msg = "Cannot open /proc/stat";
+        if (!errorContext.empty()) {
+            msg += " for " + errorContext;
+        }
+        stat_logger.error(msg);
+        return false;
+    }
+
+    std::string line;
+
+    // Skip the first "cpu" aggregate line
+    if (!std::getline(file, line)) {
+        return true;  // Nothing else to process
+    }
+
+    // Read per-CPU lines: cpu0, cpu1, cpu2, ...
+    while (std::getline(file, line)) {
+        if (line.rfind("cpu", 0) != 0 || !isdigit(line[3])) {
+            break;  // No more cpuN lines
+        }
+
+        std::istringstream iss(line);
+
+        std::string cpuLabel;
+        iss >> cpuLabel;  // e.g. "cpu0"
+
+        std::vector<long> cpuStats;
+        cpuStats.reserve(8);
+
+        long value;
+        for (int i = 0; i < 8; ++i) {
+            if (!(iss >> value)) {
+                value = 0;  // Default fallback if parsing fails
+            }
+            cpuStats.push_back(value);
         }
+
+        readings.push_back(std::move(cpuStats));
     }
+
+    return true;
+}
+
+static void getCpuCycles(long *totalp, long *idlep) {
```

Comment:
Can you see if you can use the following function instead of current getCpuCycles().

```
static void getCpuCycles(long *totalp, long *idlep) {
    *totalp = 0;
    *idlep = 0;

    std::ifstream file("/proc/stat");
    if (!file.is_open()) return;

    std::string cpu;
    long user, nice, system, idle;
    long iowait, irq, softirq, steal, guest, guest_nice;

    file >> cpu;
    if (cpu != "cpu") return;

    if (!(file >> user >> nice >> system >> idle
               >> iowait >> irq >> softirq >> steal
               >> guest >> guest_nice)) {
        return;
    }

    long total = user + nice + system + idle +
                 iowait + irq + softirq + steal +
                 guest + guest_nice;

    *totalp = total;
    *idlep = idle;
}

```